labs/a10.md

```diff
@@ -0,0 +1,371 @@
+
+Copy the contents of the Dockerfile above into a file named `Dockerfile`.
+Then use Docker to build it with the following command:
+
+`docker build -t mypython:latest .`
+
+This tells Docker to look in the current directory for a Dockerfile to build, and build it.
+The `-t` flag tells it to tag this build with the name `mypython:latest`.
+Docker will look for a Dockerfile in the current directory since you specified `.`
+
+Remember, you can see all of the images you've built on your machine with the `docker images` command.
+
+#### Questions
+1. Run the image you just built. Since we specified the default `CMD`, you can just do `docker run -it
+mypython:latest`. What do you observe?
+2. Write and build a Dockerfile that installs the packages `fortune` and `fortunes-min` and runs the `fortune`
+executable (located in `/usr/games/fortune` after you install it). Note that you won't need to use the `-it` flags when you run
+the container as `fortune` doesn't need `STDIN`.
+Submit your Dockerfile with this lab. Hint: if you're having trouble writing your Dockerfile, try booting an
+interactive container and installing both packages. Translate what you did into a Dockerfile.
+How can you translate what you did interactively to a Dockerfile?
+3. Paste the output of your `docker images` command after questions 1 and 2.
+
+### Detached Containers and Ports
+You might not always want containers to be running interactively. For instance, if you are running a web server, you'll
+likely want the container to continue keep running until you explicitly want to end it. Docker supports this use case
+with the `-d` flag, which starts containers in [detached
+mode](https://docs.docker.com/engine/reference/run/#detached--d).
+
+We'll explore a bit about detached containers by running a standalone Apache container. The image has already been
+built for you; you can find it on [DockerHub](https://hub.docker.com/_/httpd/) as `httpd`.
+
+Docker creates a separate virtual network for containers, so you will need to do forward your host port to your
+container's port (this is called [port forwarding](https://en.wikipedia.org/wiki/Port_forwarding), or port mapping).
+The container is listening on port 80, so let's try to forward our host machine's port 5050 to the container's
+port 80 when we run the container:
+
+`docker run -d -p=5050:80 httpd`
+
+The `-p` flag takes in a colon separated pair of `HOST_PORT:CONTAINER_PORT` (it can actually accept a ton of more
+options, but don't worry about that for now). You should be able to view visit `<url_of_host_machine>:5050`, assuming
+you don't have anything else running on that port (if you're not on campus, you can just `curl <url_of_host_machine>:5
+050`
+from your VM or another machine on campus, e.g. ssh.ocf.berkeley.edu), and see the words "**It works!**. You may need to allow
+the port 5050 on your firewall, simply run the command `ufw allow 5050`
```

Comment:
Nice, thanks for including that, I totally forgot when I did my lab that this was set up in a4.